home_audit.ts

```diff
@@ -0,0 +1,691 @@
+        const {foundation_type, foundation_insulation_level} = floor;
+        if(!nullOrUndefined.includes(foundation_type) && !nullOrUndefined.includes(foundation_insulation_level)) {
+            if(foundation_type === 'slab_on_grade' && ![0, 5].includes(foundation_insulation_level)) {
+                addErrorMessage(`/zone/zone_floor/${index}/foundation_insulation_level`, 'Insulation must be R-0 or R-5 for Slab on Grade Foundation');
+            } else if(![0, 11, 19].includes(foundation_insulation_level)) {
+                addErrorMessage(`/zone/zone_floor/${index}/foundation_insulation_level`, 'Insulation must be R-0, R-11, or R-19 for current foundation type');
+            }
+        }
+    });
+}
+
+/**
+ * Check that the conditioned area is within the bounds for the building footprint
+ */
+function checkConditionedAreaValid(combined_area, conditioned_footprint, area_type) {
+    const min = conditioned_footprint * 0.95;
+    const max = conditioned_footprint * 2.5;
+    if(!((min < combined_area) && (combined_area < max))) {
+        return `This home's minimum footprint is approximately ${conditioned_footprint}sqft, but you have specified ${combined_area}sqft of total ${area_type} area. The allowed range is (${Math.ceil(min)}sqft - ${Math.floor(max)}sqft). Please adjust any incorrect values. *The footprint is calculated as (<total area> - <conditioned basement area>) / <number of floors>`;
+    }
+}
+
+/**
+ * Do the additional validations for the zone floors
+ */
+function getAdditionalFloorZoneValidations(zone, about) {
+    const conditioned_footprint = getBuildingConditionedFootprint(about, zone);
+
+    // Conditioned Footprint for home must be greater than 250 sq ft.
+    if(conditioned_footprint < 250) {
+        addErrorMessage('/about/conditioned_floor_area', `Home footprint must be greater than 250 sq ft. Current footprint is ${conditioned_footprint} sq ft.`);
+    }
+
+    // Floor area is within bounds of conditioned floor area
+    checkFloorArea(zone, conditioned_footprint);
+
+    // Validate foundation insulation level is correct for foundation type
+    checkFoundationLevel(zone.zone_floor);
+}
+
+/**
+ * Helper to get the conditioned area of a particular field for a building
+ */
+function getCombinedArea(array_obj, field_name) {
+    let combined_area = 0;
+    array_obj.filter((obj) => (!nullOrUndefined.includes(obj))).forEach((obj) => {
+        if(!nullOrUndefined.includes(obj[field_name])) {
+            combined_area += obj[field_name]
+        }
+    });
+    return Math.floor(combined_area);
+}
+
+function getCombinedFloorArea(zone_floor_array) {
+    return getCombinedArea(zone_floor_array, 'floor_area');
+}
+
+function getCombinedCeilingArea(zone_roof_array) {
+    return getCombinedArea(zone_roof_array, 'ceiling_area');
+}
+
+function getCombinedRoofArea(zone_roof_array) {
+    return getCombinedArea(zone_roof_array, 'roof_area');
+}
+
+function getCombinedRoofCeilingArea(zone_roof_array) {
```

Comment:
Can do later, but I like when wrapped/derived commands use underscores to indicate (`getCombinedArea_ceiling_area`)

Reply:
This is super easy to do (shift+F6 if you don't already know, lets you rename a function/class/variable and it renames all references to it throughout the project), so I just did it.